### PR TITLE
ServiceNow Plugins JUnits

### PR DESCRIPTION
**Pull request by sgarg-CS**
[CEC-527](https://cloudsufi.atlassian.net/browse/CEC-527)

This PR has been raised for internal review to be done by the CLOUDSUFI team members.

Please leave a comment here when you've reviewed all the changes in this PR.

### CODE REVIEW

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputSplitTest.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    serviceNowInputSplit.readFields(objectInputStream);
+    verify(objectInputStream).readInt();
+    verify(objectInputStream).readUTF();
+    assertEquals("Utf", serviceNowInputSplit.getTableName());
+    assertEquals(1, serviceNowInputSplit.getOffset());
+  }
+
+  @Test
+  public void testGetLocations() throws IOException, InterruptedException {
+    assertEquals(0, (new ServiceNowInputSplit("Table Name", 2)).getLocations().length);
```
Comment (dhiraj0116) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputSplitTest.java]: Seeing the name "getLocations" looks this test should be validating string value.

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormatTest.java [hunk 2]
```diff
@@ -17,79 +17,180 @@
...
+    PowerMockito.mockStatic(HttpClientBuilder.class);
+    PowerMockito.mockStatic(RestAPIResponse.class);
+    PowerMockito.when(HttpClientBuilder.create()).thenReturn(httpClientBuilder);
+    Mockito.when(httpClientBuilder.build()).thenReturn(httpClient);
+    CloseableHttpResponse httpResponse = Mockito.mock(CloseableHttpResponse.class);
+    Mockito.when(httpClient.execute(Mockito.any())).thenReturn(httpResponse);
+    PowerMockito.when(RestAPIResponse.parse(ArgumentMatchers.any(), ArgumentMatchers.anyString())).
+            thenReturn(response);
+
+    assertEquals(0, ServiceNowInputFormat.fetchTableInfo(mode, config).size());
```
Comment (dhiraj0116) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormatTest.java]: Didn't understand what we are validating here, after doing this much mocking we are validating 0 seems bit weird.

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormatTest.java [hunk 3]
```diff
@@ -17,79 +17,180 @@
...
```
Comment (dhiraj0116) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormatTest.java]: Same comment as above

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowMultiSourceTest.java [hunk 4]
```diff
@@ -16,147 +16,212 @@
...
+      "            \"time_zone\": \"\",\n" +
+      "            \"schedule\": \"\",\n" +
+      "            \"date_format\": \"\",\n" +
+      "            \"location\": \"\"]\n" +
+      "}";
+    RestAPIResponse restAPIResponse = new RestAPIResponse(httpStatus, headers, responseBody);
+    Mockito.when(restApi.executeGet(Mockito.any())).thenReturn(restAPIResponse);
+    Mockito.when(restApi.parseResponseToResultListOfMap(restAPIResponse.getResponseBody())).thenReturn(result);
+    serviceNowMultiSource.configurePipeline(mockPipelineConfigurer);
+    Assert.assertEquals(0, mockFailureCollector.getValidationFailures().size());
```
Comment (dhiraj0116) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowMultiSourceTest.java:158]: Same comment

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowMultiSourceTest.java [hunk 5]
```diff
@@ -16,147 +16,212 @@
...
+    PowerMockito.mockStatic(HttpClientBuilder.class);
+    PowerMockito.mockStatic(RestAPIResponse.class);
+    PowerMockito.when(HttpClientBuilder.create()).thenReturn(httpClientBuilder);
+    Mockito.when(httpClientBuilder.build()).thenReturn(httpClient);
+    CloseableHttpResponse httpResponse = Mockito.mock(CloseableHttpResponse.class);
+    Mockito.when(httpClient.execute(Mockito.any())).thenReturn(httpResponse);
+    PowerMockito.when(RestAPIResponse.parse(ArgumentMatchers.any(), ArgumentMatchers.anyString())).
+      thenReturn(response);
     serviceNowMultiSource.prepareRun(context);
     Assert.assertEquals(0, mockFailureCollector.getValidationFailures().size());
```
Comment (dhiraj0116) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowMultiSourceTest.java]: From past google reviews I have seen that it's better to provide 3 parameter in assertEquals, "Failure msg", "expected value", "actual value"

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java [hunk 6]
```diff
@@ -16,108 +16,234 @@
...
+    PowerMockito.mockStatic(HttpClientBuilder.class);
+    PowerMockito.mockStatic(RestAPIResponse.class);
+    PowerMockito.when(HttpClientBuilder.create()).thenReturn(httpClientBuilder);
+    Mockito.when(httpClientBuilder.build()).thenReturn(httpClient);
+    CloseableHttpResponse httpResponse = Mockito.mock(CloseableHttpResponse.class);
+    Mockito.when(httpClient.execute(Mockito.any())).thenReturn(httpResponse);
+    PowerMockito.when(RestAPIResponse.parse(ArgumentMatchers.any(), ArgumentMatchers.anyString())).
+      thenReturn(response);
+    serviceNowSource.configurePipeline(mockPipelineConfigurer);
+    Assert.assertEquals(0, mockFailureCollector.getValidationFailures().size());
```
Comment (dhiraj0116) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java:180]: Try to validate something positive, validating failure size is not a good practice.

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java [hunk 7]
```diff
@@ -16,108 +16,234 @@
...
+    Mockito.when(restApi.executeGet(Mockito.any())).thenReturn(restAPIResponse);
+    Mockito.when(restApi.parseResponseToResultListOfMap(restAPIResponse.getResponseBody())).thenReturn(result);
     try {
       serviceNowSource.configurePipeline(mockPipelineConfigurer);
       Assert.fail("Exception is not thrown for Non-Empty Tables");
     } catch (ValidationException e) {
-      Assert.assertEquals(0, mockFailureCollector.getValidationFailures().size());
-      Assert.assertEquals("Errors were encountered during validation.",
-                          e.getMessage());
+      Assert.assertEquals("Table: sys_user is empty.", e.getFailures().get(0).getMessage());
```
Comment (dhiraj0116) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java:204]: This looks good, this is what I had mentioned above...

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java [hunk 8]
```diff
@@ -16,108 +16,234 @@
...
+    PowerMockito.mockStatic(HttpClientBuilder.class);
+    PowerMockito.mockStatic(RestAPIResponse.class);
+    PowerMockito.when(HttpClientBuilder.create()).thenReturn(httpClientBuilder);
+    Mockito.when(httpClientBuilder.build()).thenReturn(httpClient);
+    CloseableHttpResponse httpResponse = Mockito.mock(CloseableHttpResponse.class);
+    Mockito.when(httpClient.execute(Mockito.any())).thenReturn(httpResponse);
+    PowerMockito.when(RestAPIResponse.parse(ArgumentMatchers.any(), ArgumentMatchers.anyString())).
+      thenReturn(response);
     serviceNowSource.prepareRun(context);
     Assert.assertEquals(0, mockFailureCollector.getValidationFailures().size());
```
Comment (dhiraj0116) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java:309]: Same comment as above

Comment (sgarg-CS) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java:309]: Added another assertion for checking whether the `context` has a non-null `multisink.table name` argument value. c0e6a0f

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java (hunk 6 above)
Comment (sgarg-CS) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceTest.java:180]: Added another assertion for checking `Schema.Type` c0e6a0f

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowMultiSourceTest.java (hunk 5 above)
Comment (sgarg-CS) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowMultiSourceTest.java]: Added the failure message. c0e6a0f

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowMultiSourceTest.java (hunk 4 above)
Comment (sgarg-CS) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowMultiSourceTest.java:158]: Added the assertion for output schema as well. c0e6a0f

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormatTest.java (hunk 3 above)
Comment (sgarg-CS) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormatTest.java]: Added the new result and updated the assertion. c0e6a0f

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormatTest.java (hunk 2 above)
Comment (sgarg-CS) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormatTest.java]: I've replaced the empty result with some data and updated the assertion accordingly. c0e6a0f

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputSplitTest.java (hunk 1 above)
Comment (sgarg-CS) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowInputSplitTest.java]: Added the assertion for checking the class type. c0e6a0f

**Comment by dhiraj0116** [comment]
Looks good, after the changes...

**Pull request merged**